stock_picking_propagate_scheduled_date/models/stock_move.py

```diff
@@ -2,32 +2,80 @@
-                move_dest.date -= delta
-
     def write(self, vals):
-        # propagate date changes in the stock move chain
-        if "date" in vals:
-            self._propagate_date(vals.get("date"))
-        return super().write(vals)
+        """Override the write method to propagate the date to the destination moves."""
+        need_propagate = "date" in vals and not self.env.context.get(
+            "_propagate_date_no_recursion"
+        )
+        if need_propagate:
+            # Get the previous date of the moves
+            previous_dates = {
+                move.id: move.date for move in self.with_context(prefetch_fields=False)
+            }
+        res = super().write(vals)
+        if need_propagate:
+            self._propagate_date_no_recursion(previous_dates, vals.get("date"))
+        return res
+
+    def _get_moves_in_down_chain(self):
+        """Get all the destination moves in the down chain of a moves recordset recursively."""
+        query = """
+            WITH RECURSIVE chained_moves AS (
+                SELECT
+                    r.move_orig_id, r.move_orig_id as from_id, r.move_dest_id as dest_id
+                FROM
+                    stock_move_move_rel r
+                WHERE r.move_orig_id IN %s
+
+                UNION ALL
+
+                SELECT
+                    chained_moves.move_orig_id,
+                    r.move_orig_id as from_id,
+                    r.move_dest_id as dest_id
+                FROM
+                    stock_move_move_rel r
+                JOIN chained_moves ON r.move_orig_id = chained_moves.dest_id
+            )
+
+            SELECT
+                move_orig_id,
+                array_agg(dest_id)
+            FROM chained_moves
+            GROUP BY move_orig_id;
+        """
+
+        self.env.cr.execute(query, (tuple(self.ids),))
+        # key: original move, values: list of destination moves
+        chained = dict(self.env.cr.fetchall())
+        return chained
+
+    def _propagate_date_no_recursion(self, previous_dates, new_date):
+        """Propagate the date of a move to its destination avoiding recursion."""
+        if not self:
+            return True
+        chained_dest_moves = self._get_moves_in_down_chain()
+        if chained_dest_moves:
+            for move in self:
+                prev_date = previous_dates.get(move.id)
+                dest_moves = set(chained_dest_moves.get(move.id))
+                if prev_date:
+                    delta = prev_date - fields.Datetime.to_datetime(new_date)
+                else:
+                    continue
+                for move_dest in self.browse(dest_moves):
+                    if move_dest.state in ("done", "cancel"):
+                        continue
+                    move_dest.with_context(_propagate_date_no_recursion=True).write(
+                        {
+                            "date": move_dest.date - delta,
+                        }
+                    )
```

Comment:
```suggestion
                dest_move_ids = chained_dest_moves.get(move.id)
                if not dest_move_ids:
                    continue
                prev_date = previous_dates.get(move.id)
                if not prev_date:
                    continue
                delta = prev_date - fields.Datetime.to_datetime(new_date)
                dest_moves = self.browse(set(dest_move_ids))
                for dest_move in dest_moves.with_context(_propagate_date_no_recursion=1):
                    if dest_move.state not in ("done", "cancel"):
                        dest_move.date = move_dest.date - delta
```

Reply:
If we inject context on the recordset we get into recursion. So, the correct way is to inject by record.

Reply:
Having
```
for rec in self:
    rec.with_context(flag=value).do_something()
```
is perfectly equivalent to
```
for rec in self.with_context(flag=value):
    rec.do_something()
```
The only difference is that, in the second case, you update the context on the whole recordset at once, instead of updating the context for each record singularly.

Reply:
I understand that. I'll check again.